// @covalent/core/layout/layout-nav/layout-nav.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component } from '@angular/core';
import { Input, Output, EventEmitter } from '@angular/core';
import { TdLayoutService } from '../services/layout.service';
export var TdLayoutNavComponent = (function () {
    function TdLayoutNavComponent(layoutService) {
        this.layoutService = layoutService;
        /**
         * method thats called when menu is clicked
         */
        this.onOpenMenu = new EventEmitter();
    }
    TdLayoutNavComponent.prototype.menuClick = function () {
        this._onMenuClick();
    };
    /**
     * emits menuEvent
     */
    TdLayoutNavComponent.prototype._onMenuClick = function () {
        this.onOpenMenu.emit(undefined);
        this.layoutService.openSideNav('menu');
    };
    __decorate([
        Input('title'), 
        __metadata('design:type', String)
    ], TdLayoutNavComponent.prototype, "title", void 0);
    __decorate([
        Input('icon'), 
        __metadata('design:type', String)
    ], TdLayoutNavComponent.prototype, "icon", void 0);
    __decorate([
        Input('logo'), 
        __metadata('design:type', String)
    ], TdLayoutNavComponent.prototype, "logo", void 0);
    __decorate([
        Output('openMenu'), 
        __metadata('design:type', EventEmitter)
    ], TdLayoutNavComponent.prototype, "onOpenMenu", void 0);
    TdLayoutNavComponent = __decorate([
        Component({
            selector: 'td-layout-nav',
            styles: [""],
            template: "<div layout=\"column\" layout-fill> <md-toolbar color=\"primary\"> <button md-button (click)=\"menuClick()\" md-icon-button><md-icon class=\"md-24\">menu</md-icon></button> <md-icon *ngIf=\"icon\">{{icon}}</md-icon> <md-icon *ngIf=\"logo\" class=\"md-icon-logo\" svgSrc=\"{{logo}}\"></md-icon> <span>{{title}}</span> <ng-content select=\"[toolbar-content]\"></ng-content> </md-toolbar> <div flex class=\"content md-content\"> <ng-content></ng-content> </div> </div> ",
        }), 
        __metadata('design:paramtypes', [TdLayoutService])
    ], TdLayoutNavComponent);
    return TdLayoutNavComponent;
}());
