// @covalent/file-upload/file-upload.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpModule, JsonpModule } from '@angular/http';
import { FormsModule } from '@angular/forms';
import { MaterialModule } from '@angular/material';
import { TdFileSelectDirective } from './directives/file-select.directive';
import { TdFileDropDirective } from './directives/file-drop.directive';
import { TdFileUploadComponent } from './file-upload.component';
import { TdFileService } from './services/file.service';
export var TD_FILE_DIRECTIVES = [
    TdFileSelectDirective,
    TdFileDropDirective,
    TdFileUploadComponent,
];
export var CovalentFileModule = (function () {
    function CovalentFileModule() {
    }
    CovalentFileModule.forRoot = function () {
        return {
            ngModule: CovalentFileModule,
            providers: [TdFileService],
        };
    };
    CovalentFileModule = __decorate([
        NgModule({
            imports: [
                HttpModule,
                JsonpModule,
                FormsModule,
                CommonModule,
                MaterialModule.forRoot(),
            ],
            declarations: [
                TD_FILE_DIRECTIVES,
            ],
            exports: [
                TD_FILE_DIRECTIVES,
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], CovalentFileModule);
    return CovalentFileModule;
}());
