// @covalent/core/layout/layout-nav-list/layout-nav-list.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component } from '@angular/core';
import { Input, Output, EventEmitter } from '@angular/core';
import { ViewChild } from '@angular/core';
import { MdSidenav } from '@angular/material';
import { TdLayoutService } from '../services/layout.service';
export var TdLayoutNavListComponent = (function () {
    function TdLayoutNavListComponent(layoutService) {
        this.layoutService = layoutService;
        this._transitioning = false;
        /**
         * method thats called when menu is clicked
         */
        this.onOpenMenu = new EventEmitter();
    }
    TdLayoutNavListComponent.prototype.menuClick = function () {
        this._onMenuClick();
    };
    /**
     * Proxy toggle method to access sidenav from outside (from td-layout template).
     */
    TdLayoutNavListComponent.prototype.toggle = function () {
        var _this = this;
        if (!this._transitioning) {
            this._transitioning = true;
            this._sideNav.toggle().then(function () {
                _this._transitioning = false;
            });
        }
    };
    /**
     * Proxy open method to access sidenav from outside (from td-layout template).
     */
    TdLayoutNavListComponent.prototype.open = function () {
        var _this = this;
        if (!this._transitioning) {
            this._transitioning = true;
            this._sideNav.open().then(function () {
                _this._transitioning = false;
            });
        }
    };
    /**
     * Proxy close method to access sidenav from outside (from td-layout template).
     */
    TdLayoutNavListComponent.prototype.close = function () {
        var _this = this;
        if (!this._transitioning) {
            this._transitioning = true;
            this._sideNav.close().then(function () {
                _this._transitioning = false;
            });
        }
    };
    /**
     * emits menuEvent
     */
    TdLayoutNavListComponent.prototype._onMenuClick = function () {
        this.onOpenMenu.emit(undefined);
        this.layoutService.openSideNav('menu');
    };
    __decorate([
        Input('title'), 
        __metadata('design:type', String)
    ], TdLayoutNavListComponent.prototype, "title", void 0);
    __decorate([
        Input('icon'), 
        __metadata('design:type', String)
    ], TdLayoutNavListComponent.prototype, "icon", void 0);
    __decorate([
        Input('logo'), 
        __metadata('design:type', String)
    ], TdLayoutNavListComponent.prototype, "logo", void 0);
    __decorate([
        ViewChild(MdSidenav), 
        __metadata('design:type', MdSidenav)
    ], TdLayoutNavListComponent.prototype, "_sideNav", void 0);
    __decorate([
        Output('openMenu'), 
        __metadata('design:type', EventEmitter)
    ], TdLayoutNavListComponent.prototype, "onOpenMenu", void 0);
    TdLayoutNavListComponent = __decorate([
        Component({
            selector: 'td-layout-nav-list',
            styles: ["/** * Mixin that creates a new stacking context. * see https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context */ :host /deep/ md-sidenav-layout > .md-sidenav-content { margin-left: 0 !important; } @media (min-width: 600px) { :host /deep/ md-sidenav-layout { height: 100%; display: flex; } :host /deep/ md-sidenav-layout > md-sidenav { display: flex; position: relative; height: 100%; width: 382px; flex: 0 0 auto; transition: none; } :host /deep/ md-sidenav-layout > md-sidenav.md-sidenav-closed, :host /deep/ md-sidenav-layout > md-sidenav.md-sidenav-closing { visibility: visible; transform: translate3d(0, 0, 0); } :host /deep/ md-sidenav-layout > .md-sidenav-content { margin-left: 0 !important; display: flex; flex-grow: 1; } } @media (max-width: 599px) { :host /deep/ md-sidenav-layout > md-sidenav.md-sidenav-opened, :host /deep/ md-sidenav-layout > md-sidenav.md-sidenav-opening, :host /deep/ md-sidenav-layout > md-sidenav.md-sidenav-closing { z-index: 3; } :host /deep/ md-sidenav-layout > .md-sidenav-content { z-index: 2; } } md-toolbar { box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26); z-index: 1; } md-toolbar.md-primary { color: white; } .md-sidenav-left { z-index: 1; } .list { background-color: #fff; text-align: start; } "],
            template: "<div layout=\"column\" layout-fill> <div class=\"md-content\" flex layout-fill> <md-sidenav-layout layout=\"row\" layout-fill> <md-sidenav opened align=\"start\" mode=\"side\" layout=\"column\" layout-fill class=\"md-sidenav-left app-list md-whiteframe-z1\" flex-gt-xs=\"none\"> <md-toolbar color=\"primary\" class=\"md-whiteframe-z1\"> <button md-button (click)=\"menuClick()\" md-icon-button><md-icon class=\"md-24\">menu</md-icon></button> <md-icon *ngIf=\"icon\">{{icon}}</md-icon> <md-icon *ngIf=\"logo\" class=\"md-icon-logo\" svgSrc=\"{{logo}}\"></md-icon> <span>{{title}}</span> <ng-content select=\"[list-toolbar-content]\"></ng-content> </md-toolbar> <div flex class=\"list md-content\"> <ng-content select=\"[list-items]\"></ng-content> </div> </md-sidenav> <div layout=\"column\" layout-fill class=\"md-content content\"> <md-toolbar color=\"primary\" class=\"md-whiteframe-z1\"> <button md-button (click)=\"toggle()\" md-icon-button hide-gt-xs><md-icon class=\"md-24\">arrow_back</md-icon></button> <ng-content select=\"[nav-toolbar-content]\"></ng-content> </md-toolbar> <div class=\"md-content\" flex> <ng-content></ng-content> </div> </div> </md-sidenav-layout> </div> </div> ",
        }), 
        __metadata('design:paramtypes', [TdLayoutService])
    ], TdLayoutNavListComponent);
    return TdLayoutNavListComponent;
}());
