// @covalent/highlight/highlight.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global.td = global.td || {}, global.td.highlight = global.td.highlight || {}),global.ng.core,global.ng.common));
}(this, (function (exports,_angular_core,_angular_common) { 'use strict';

var __decorate = (window && window.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (window && window.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
/* tslint:disable-next-line */
var hljs = require('highlight.js/lib');
var TdHighlightComponent = (function () {
    function TdHighlightComponent(renderer) {
        this.renderer = renderer;
        this.language = 'javascript';
    }
    TdHighlightComponent.prototype.ngAfterViewInit = function () {
        if (!this.language) {
            throw 'Error: language attribute must be defined in TdHighlightComponent.';
        }
        var codeElement = this.content.nativeElement;
        var code = codeElement.innerHTML;
        this.renderer.detachView([].slice.call(codeElement.childNodes));
        this.render(code, codeElement);
    };
    TdHighlightComponent.prototype.render = function (contents, codeElement) {
        var lines = contents.split('\n');
        // Remove empty lines
        lines = lines.filter(function (line) {
            return line.trim().length > 0;
        });
        // Make it so each line starts at 0 whitespace
        var firstLineWhitespace = lines[0].match(/^\s*/)[0];
        var startingWhitespaceRegex = new RegExp('^' + firstLineWhitespace);
        lines = lines.map(function (line) {
            return line
                .replace('=""', '') // remove empty values
                .replace(startingWhitespaceRegex, '')
                .replace(/\s+$/, '');
        });
        this.renderer.setElementClass(codeElement, 'highlight', true);
        var codeToParse = lines.join('\n')
            .replace(/\{ \{/gi, '{{').replace(/\} \}/gi, '}}')
            .replace(/&lt;/gi, '<').replace(/&gt;/gi, '>'); // replace with < and > to render HTML in angular 2
        if (this.language === 'html') {
            this.renderer.createText(codeElement, codeToParse, undefined);
            hljs.highlightBlock(codeElement);
        }
        else {
            var highlightedCode = hljs.highlight(this.language, codeToParse, true);
            highlightedCode.value = highlightedCode.value
                .replace(/=<span class="hljs-value">""<\/span>/gi, '')
                .replace('<head>', '')
                .replace('<head/>', '');
            codeElement.innerHTML = highlightedCode.value;
        }
    };
    __decorate([
        _angular_core.Input('lang'), 
        __metadata('design:type', String)
    ], TdHighlightComponent.prototype, "language", void 0);
    __decorate([
        _angular_core.ViewChild('highlight'), 
        __metadata('design:type', _angular_core.ElementRef)
    ], TdHighlightComponent.prototype, "content", void 0);
    TdHighlightComponent = __decorate([
        _angular_core.Component({
            selector: 'td-highlight',
            styles: ["pre, code, .highlight { font-family: Menlo, Monaco, \"Andale Mono\", \"lucida console\", \"Courier New\", monospace; } pre { display: block; overflow-x: auto; padding: 0; margin: 0; background: transparent; color: white; font-family: Menlo, Monaco, \"Andale Mono\", \"lucida console\", \"Courier New\", monospace; line-height: 1.45; tab-size: 2; -webkit-font-smoothing: auto; -webkit-text-size-adjust: none; position: relative; border-radius: 2px; font-size: 0.8rem; } code { margin: 0; padding: 0; overflow-wrap: break-word; white-space: pre-wrap; } .highlight { display: block; overflow-wrap: break-word; color: #e8eaf6; line-height: 1.5; margin: 0; } /* Use host as a parent here since hljs renders after dom & doesn't get Angular shadowdom */ :host { display: block; overflow-x: auto; background: #292348; color: white; padding: 16px; } :host + :host { margin-top: 16px; /* 2nd neighbor & beyond will have margin */ } :host /deep/ .hljs-comment, :host /deep/ .hljs-quote { color: #ce93d8; } :host /deep/ .hljs-variable, :host /deep/ .hljs-template-variable, :host /deep/ .hljs-tag, :host /deep/ .hljs-name, :host /deep/ .hljs-selector-id, :host /deep/ .hljs-selector-class, :host /deep/ .hljs-regexp, :host /deep/ .hljs-deletion { color: #ff80ab; } :host /deep/ .hljs-number, :host /deep/ .hljs-built_in, :host /deep/ .hljs-builtin-name, :host /deep/ .hljs-literal, :host /deep/ .hljs-type, :host /deep/ .hljs-params, :host /deep/ .hljs-meta, :host /deep/ .hljs-link { color: #ffcc80; } :host /deep/ .hljs-attribute { color: #ffab91; } :host /deep/ .hljs-string, :host /deep/ .hljs-symbol, :host /deep/ .hljs-bullet, :host /deep/ .hljs-addition { color: #e6ee9c; } :host /deep/ .hljs-title, :host /deep/ .hljs-section { color: #81d4fa; } :host /deep/ .hljs-keyword { color: #64ffda; } :host /deep/ .hljs-selector-tag { color: #82b1ff; } :host /deep/ .hljs-emphasis { font-style: italic; } :host /deep/ .hljs-strong { font-weight: bold; } "],
            template: "<pre> <code #highlight> <ng-content></ng-content> </code> </pre>",
        }), 
        __metadata('design:paramtypes', [_angular_core.Renderer])
    ], TdHighlightComponent);
    return TdHighlightComponent;
}());

var __decorate$1 = (window && window.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata$1 = (window && window.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var CovalentHighlightModule = (function () {
    function CovalentHighlightModule() {
    }
    CovalentHighlightModule.forRoot = function () {
        return {
            ngModule: CovalentHighlightModule,
            providers: [],
        };
    };
    CovalentHighlightModule = __decorate$1([
        _angular_core.NgModule({
            imports: [
                _angular_common.CommonModule,
            ],
            declarations: [
                TdHighlightComponent,
            ],
            exports: [
                TdHighlightComponent,
            ],
        }), 
        __metadata$1('design:paramtypes', [])
    ], CovalentHighlightModule);
    return CovalentHighlightModule;
}());

exports.TdHighlightComponent = TdHighlightComponent;
exports.CovalentHighlightModule = CovalentHighlightModule;

Object.defineProperty(exports, '__esModule', { value: true });

})));
