// @covalent/data-table/data-table.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { MaterialModule } from '@angular/material';
import { TdDataTableComponent } from './data-table.component';
import { TdDataTableTableDirective } from './directives/table.directive';
import { TdDataTableHeadDirective } from './directives/head.directive';
import { TdDataTableBodyDirective } from './directives/body.directive';
import { TdDataTableRowDirective } from './directives/row.directive';
import { TdDataTableColumnDirective } from './directives/column.directive';
import { TdDataTableCellDirective } from './directives/cell.directive';
export var TD_DATA_TABLE_DIRECTIVES = [
    TdDataTableComponent,
    TdDataTableTableDirective,
    TdDataTableHeadDirective,
    TdDataTableBodyDirective,
    TdDataTableRowDirective,
    TdDataTableColumnDirective,
    TdDataTableCellDirective,
];
export var CovalentDataTableModule = (function () {
    function CovalentDataTableModule() {
    }
    CovalentDataTableModule.forRoot = function () {
        return {
            ngModule: CovalentDataTableModule,
            providers: [],
        };
    };
    CovalentDataTableModule = __decorate([
        NgModule({
            imports: [
                MaterialModule.forRoot(),
                CommonModule,
                FormsModule,
                ReactiveFormsModule,
            ],
            declarations: [
                TD_DATA_TABLE_DIRECTIVES,
            ],
            exports: [
                TD_DATA_TABLE_DIRECTIVES,
            ],
        }), 
        __metadata('design:paramtypes', [])
    ], CovalentDataTableModule);
    return CovalentDataTableModule;
}());
